mne/io/boxy/boxy.py

```diff
@@ -0,0 +1,301 @@
+# Authors: Kyle Mathewson, Jonathan Kuziek <kuziek@ualberta.ca>
+#
+# License: BSD (3-clause)
+
+import re as re
+
+import numpy as np
+
+from ..base import BaseRaw
+from ..meas_info import create_info
+from ...utils import logger, verbose, fill_doc
+from ...annotations import Annotations
+
+
+@fill_doc
+def read_raw_boxy(fname, datatype='AC', preload=False, verbose=None):
+    """Reader for a BOXY optical imaging recording.
+
+    Parameters
+    ----------
+    fname : str
+        Path to the BOXY data file.
+    datatype : str
+        Type of data to return (AC, DC, or Ph).
```

Comment:
Would it be better to always give all three? For example the first three MNE channels are AC/DC/Ph of the first optode, next three MNE channels are from the second optode, etc.?

Reply:
the data size gets very large if you pick all three, channel count is very high in modern systems (16 detectors X 16 sources x 2), I think this could introduce a problem, 

also usually one would only be analyzing one of these three types (to either do a slow NIRS analysis or a fast EROS analysis)

Reply:
What is eros?

Reply:
eros stands for event-related optical signal , used in the literature to describe averaging multiple short trials of phase delay data in a frequency domain system, I don't like this name because slower NIRS signals are also trial averaged as in your tutorials. So I prefer to use fast signal and slow signal

the fast signal is a decrease in scattering in active tissue simultaneous with electrical changes in the neurons (David Rector references). The slow signal is in increase in absorption of light due to blood 6-10 seconds following neural activity. 

normally fast signal is combined over short and long wavelengths as well, another difference, instead of converting to oxy and deoxy hemoglobin

we can bring all this up in the tutorial on fast optical analysis in MNE

Reply:
Thanks so much @kylemath, this whole PR has been really informative for me.

Reply:
> the data size gets very large if you pick all three

Do you actually load all possible (16x16x2) combinations of sources and detectors? Or only the ones with distances feasible for the light to travel over?

Even still, how much data are we talking about? I don't have a gut feeling for this because CW nirs has such a slow sample rate, but I think FD has a faster sample rate. Are we talking more than a GB of data per hour? What is a typical measurement data size?

Reply:
> the data size gets very large if you pick all three, channel count is very high in modern systems (16 detectors X 16 sources x 2), I think this could introduce a problem,

It shouldn't based on how `_read_segment_file` and `BaseRaw` work. We support users doing `read_raw_whatever(...).pick_types(...).load_data()`, i.e., loading data *after* picking channels to load.

If you are already coding it to know how to triage over the three channel types, it should be possible to do so in `_read_segment_file` such that, if you have `.pick_types(...)`'ed only one of the three channel types, it will end up being as fast as however you have it implemented now.

So I'd say for now implement reading all three in `_read_segment_file`, and we can make sure it does so efficiently in cases where you only want one of these channel types. The easiest way to do this is to implement the naive version of the reader first that always reads all channels, then just sub-indexes the necessary ones in `_read_segment_file`. Once that 100% works (even if it's slow) then you or I or someone else can work on making it more efficient.

So to me the efficiency question we can tackle later, and here we only need to consider whether or not the public API using `raw.pick_types(...)` makes more sense than having a parameter for the channel types to read in `read_raw_boxy`.

Reply:
I believe it selects all combinations, not just the good distance ones,
Its about 1GB per hour I believe

@kuziekj do you have an idea bout how to implement this? do you have any questions?

Reply:
1 GB per hour is less than most MEG recordings, so I don't anticipate this being a bottleneck. We should just implement a version that reads everything then try to make it performant. If we merge this PR after 0.21 releases, it gives us 6 mo to iterate on it to improve it before the next release.

Reply:
Okay, I can certainly change this to read all data types (AC, DC, Ph) since we had an earlier version that essentially returned everything (although I understand we don't necessarily want to return everything).

At this point do we care how the data would be organised? How I had this set up before, all the DC data would be first, followed by all the AC, and then the Ph.

Also, I'm thinking `raw.pick_types(...)` would be better since I imagine it would be more consistent with other readers. It would be more complicated though since I'm not entirely sure where to start, but I can take a look.

If we go the parameter/sub-index route, would we essentially leave the reader as-is and just modify it to allow combinations of data types to be returned? So for example, `read_raw_boxy(boxy_file, ['DC', 'Ph'])` would return both the DC and Ph in a single raw data object. I would be able to get something like this ready much more quickly.

 Thoughts?

Reply:
> Okay, I can certainly change this to read all data types (AC, DC, Ph) since we had an earlier version that essentially returned everything (although I understand we don't necessarily want to return everything).

Yep -- the first version of the reader can basically read everything, then just sub-index / sub-select with the given picks. It will not be efficient but it will work. Efficiency can be the next step.

> At this point do we care how the data would be organised? How I had this set up before, all the DC data would be first, followed by all the AC, and then the Ph.

My vote would be to order by channel. So the first three are all the same channel, then the next three, etc. So AC/DC/Ph changes fastest. It makes picking AC equivalent to `[::3]`, DC `[1::3]`, phase `[2::3]`. There is precedent for this for MEG as Neuromag systems have two orthogonal planar gradiometer coils and one axial magnetometer coil at each location, and they are multiplexed this way. Also I think @rob-luke ordered the source-detector pairs this way.

> If we go the parameter/sub-index route ... I would be able to get something like this ready much more quickly.

I would give the speed of implementation the lowest priority here -- I'm confident we can get the `pick` route working, so don't let the implementation challenge stop us from doing it. One of the many benefits of social coding is that you don't have to do it all yourself -- @rob-luke recently worked through the `picks` gymnastics and so did I, so I'm sure either one of us would be happy to work on it if you want help implementing it.

Reply:
Okay, sounds like a plan. I'll start working on this and I'll be sure to ask for help if I run into any significant roadblocks.

Thanks for all the help and explanations so far. Appreciate it.